include/iarc7_fc_comms/CommonFcComms.hpp

```diff
@@ -94,90 +113,165 @@ CommonFcComms<T>& CommonFcComms<T>::getInstance()
+    if (status != FcCommsReturns::kReturnOk) {
+        ROS_ERROR("Failed to retrieve flight controller status");
+        return status;
+    }
     fc.armed = temp_armed;
     fc.auto_pilot = temp_auto_pilot;
     fc.failsafe = temp_failsafe;
-    
-    std_msgs::Float32 battery;
-    #pragma GCC warning "TODO handle failure"
-    flightControlImpl_.getBattery(battery.data);
-
     ROS_DEBUG("Autopilot_enabled: %d", fc.auto_pilot);
     ROS_DEBUG("Armed: %d", fc.armed);
     status_publisher.publish(fc);
+
+    std_msgs::Float32 battery;
+    status = flightControlImpl_.getBattery(battery.data);
+    if (status != FcCommsReturns::kReturnOk) {
+        ROS_ERROR("Failed to retrieve flight controller battery info");
+        return status;
+    }
     ROS_DEBUG("Battery level: %f", battery.data);
     battery_publisher.publish(battery);
 
-
     double attitude[3];
-    flightControlImpl_.getAttitude(attitude);
+    status = flightControlImpl_.getAttitude(attitude);
+    if (status != FcCommsReturns::kReturnOk) {
+        ROS_ERROR("Failed to retrieve attitude from flight controller");
+        return status;
+    }
     ROS_DEBUG("Attitude: %f %f %f", attitude[0], attitude[1], attitude[2]);
     sendOrientationTransform(attitude);
+
+    return FcCommsReturns::kReturnOk;
+}
+
+// Attempt to reconnect, blocking until successful
+template<class T>
+void CommonFcComms<T>::reconnect(const ros::Duration& timeout) {
+    FcCommsReturns status;
+    ros::Time start_time = ros::Time::now();
+    do {
+        status = flightControlImpl_.disconnect();
+    } while (status != FcCommsReturns::kReturnOk
+          && (timeout == ros::Duration(0) || ros::Time::now() - start_time < timeout));
+
+    do {
+        status = flightControlImpl_.connect();
+    } while (status != FcCommsReturns::kReturnOk
+          && (timeout == ros::Duration(0) || ros::Time::now() - start_time < timeout));
 }
 
 // Update the sensors on the flight controller
 template<class T>
 void CommonFcComms<T>::updateSensors(const ros::TimerEvent&)
 {
-
     ros::Time times = ros::Time::now();
 
+    FcCommsReturns status;
 
     // Do different things based on the current connection status.
     switch(flightControlImpl_.getConnectionStatus())
     {
         case FcCommsStatus::kDisconnected:
             ROS_WARN("FC_Comms disconnected");
-            // We don't care about the return value we just reconnect.
-            (void)flightControlImpl_.connect();
+            reconnect();
             break;
-        
+
         case FcCommsStatus::kConnected:
-            #pragma GCC warning "TODO handle failure"
-            flightControlImpl_.handleComms();
-            publishTopics();
+            status = flightControlImpl_.handleComms();
+            while (status != FcCommsReturns::kReturnOk) {
+                reconnect();
+                status = flightControlImpl_.handleComms();
+            }
+
+            while (have_new_direction_command_message_) {
```

Comment:
I would think that running the loop at 20Hz, but processing things that aren't orientation (either commands or measurements) at like 5Hz would work ok, but definitely no guarantees.  Either way I'd like to make sure unnecessary messages aren't being sent, that's definitely a possibility.

I'm going to make a trello card to test this at different frequencies, and I don't plan to merge this until that's done.  If we find out that with that change it doesn't run reliably at any reasonable frequency then I'm guessing we'll need to do something more than refactor, so we can decide when we get there.